### DISCUSSION TITLE
Demagnetizing Field and Custom Kernel

### DISCUSSION THREAD

**Discussion by alanchavez86** [idaholab/moose#17899] [Q&A General] [answered]
Hi all!

I've asked a similar question before and received a great discussion! I wanted to run by some thoughts I had for solving for the demagnetizing field of FeCrCo. 

In the last discussion, I was given a couple equations to work with that resembled ones that I've been working with that make the problem easier to work with:

![image](https://user-images.githubusercontent.com/74686859/119027530-ae2cbb00-b963-11eb-9b8a-e31f0b11b112.png)
and
![image](https://user-images.githubusercontent.com/74686859/119027559-b553c900-b963-11eb-8ad0-6d3ef806601a.png)

I wanted to make sense of this and computed the weak form of the first equation:

![image](https://user-images.githubusercontent.com/74686859/119027988-2e532080-b964-11eb-8bff-12c8a892165d.png)

Since I'm trying to calculate for psi, how would I go about creating a custom kernel for that? If I have a custom kernel for ∇⋅M, would I combine that with a kernel for ∇^2 Ψ? Also, ∇^2 Ψ is very similar to the diffusion kernel in moose, so would it be alright to use that?

Thank you!

**Comment by alanchavez86** [comment, marked as answer]
In this case, M should be the saturation magnetization (M = M(c)). The first source I'm attaching is from Koyama 2004. I'm assuming I can implement a material property that defines M as Ms*m(r). 
[Koyama_2004.pdf](https://github.com/idaholab/moose/files/6549846/Koyama_2004.pdf)

A custom kernel was created for del M and I was wondering how that would be put together with psi in this equation: 
![image](https://user-images.githubusercontent.com/74686859/119737214-28fe4600-be3c-11eb-8991-930e75eada2b.png)

Here is the .C file that I created for del M:
#include "MagnetizationVector.h"

/**
 * All MOOSE based object classes you create must be registered using this macro.  The first
 * argument is the name of the App you entered in when running the stork.sh script with an "App"
 * suffix. If you ran "stork.sh Example", then the argument here becomes "ExampleApp". The second
 * argument is the name of the C++ class you created.
 */
registerMooseObject("fecrcoApp", MagnetizationVector);

/**
 * This function defines the valid parameters for
 * this Kernel and their default values
 */
template <>
InputParameters
validParams<MagnetizationVector>()
{
  InputParameters params = validParams<Kernel>();
  params.addRequiredParam<RealVectorValue>("mag x", "X direction Vector");
  return params;
}

MagnetizationVector::MagnetizationVector(const InputParameters & parameters)
  : // You must call the constructor of the base class first
    Kernel(parameters),
    m(getParam<RealVectorValue>("mag x"))
{
}

Real
MagnetizationVector::computeQpResidual()
{
//  RealVectorValue m(_mag_x[_qp]);
  return - m*_grad_test[_i][_qp];
}

I removed _mag_x[_qp], _mag_y[_qp], and _mag_z[_qp] because I was assuming the vector direction wouldn't matter at higher temperatures. Here is the last part of my .h file that I modified for the custom kernel to compile correctly:

private:
  /**
   * A vector object for storing the velocity.  Convenient for
   * computing dot products.
   */
    RealVectorValue m;

If my del M custom kernel is correct, could I combine that with the Diffusion kernel to satisfy the Poisson Equation I have and how would that be done? Here is another file that I'm referencing for the demagnetizing energy:
[Feng_2017_Peritectic PF Mag.pdf](https://github.com/idaholab/moose/files/6549889/Feng_2017_Peritectic.PF.Mag.pdf)

**Reply by GiudGiud** [reply]
@mangerij @cticenhour

**Reply by mangerij** [reply]
Sorry, I didn't see the first post. I was on a bit of a vacation so sorry for the delay here. 

Yeah, this should work. You can have two separate Kernels for the two terms in the Poisson equation (just make sure the signs are right!). The surface terms in the equation either drop out or are handled naturally within MOOSE. As for your pseudo code here:

Note that m*_grad_test[_i][_qp] should be a dot product because _grad_test[_i][_qp] is a vector (gradient of a scalar function). This is where Koyama2004 confuses me because he seems to be concerned with the vector value of the magnetization.

Also, the magnetization should depend explicitly on the concentration. What you could do is write an AuxKernel to calculate the magnetization m = m(c(**r**)) and then call this AuxVariable in your Kernel for grad*m. I am not sure if you need to calculate the off-diagonal terms for dm/dc in the Kernel for grad"m -  you might :)

**Reply by alanchavez86** [reply]
Thanks for the advice John! I just posted a different discussion that me and my colleague are looking into to solve for the magnetic potential! We modified the CoefDiffusion Kernel to take a function and solve for phi with good results. Now the question is how can we can add a material property that can act as the function and solve for phi.
